### PR TITLE
Adding the Quantum Oracle Tutorial

### PR DESCRIPTION
**Pull request by alexandre-ricardo**
Adding the Quantum Oracle Tutorial, with a simple example of the Deutsch-Jozsa algorithm as application.

### CODE REVIEW

File: tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,183 @@
+{
```
Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>Over all looks really good, clear and to the point.</p><p>Well done!</p><p> </p><p>Let me know if you have questions, or if you want further clarifications for any of my notes.</p><p> </p><p>If you don't agree with any of the notes it perfectly fine and we should discuss it, or just keep it as a question for the review with Eden.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>Minor typo in the last paragraph ("fre" should be three).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>I think it's a good idea to remind/ explain the reader that <code>^=</code> is bitwise XOR of the binary representation of <code>res</code>  and  &nbsp;<code>x%2 == 1</code> and assigning the result on <code>res</code>.</p><p>I would also consider explaining again shortly what <code>x %2 == 1</code> does (oddity check).</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.        res ^= x %2 == 1
</pre><p>If I understand correctly, we could have change the == 1 to ==0, and then later omit the X(y) in the main, right? (checking if the number is even instead if its odd).</p><p> </p><p>It's quite minor, mainly my OCD, but might "cleaner" and slightly clearer. I don't have to change it if you don't feel like, but if so, I just would have add clarification on the purpose of the X(y).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.        X(target=y)
</pre><p>consider omitting ("==1" -&gt; "==0")/ or explaining shortly. See note above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        apply_to_all(H,x)
</pre><p>cold this line and the two ones that comes afterwords be combined as "within_apply" statement? If so, I would have implement it this way (maybe as separate function), add to the "classiq concepts" list above with the links.Also add under the function of with the "within_apply" few word on this concept (compute and uncompute, freeing qubits for further use).</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>Maybe before "References" you might add mathematical analysis, if it's not too heavy. </p><p> </p><p>Also, Eden's example tutorial includes "All together" part of the code, and Qmod implementation (you can use write "write_qmod" function and just place it's output in markdown cell).</p><p> </p><p>Also consider adding executing the oracle and showing that the results are consist with our first intention (could be done from IDE "show" "execute" or directly from the SDK, for your choice). </p><p> </p><p>Check this for references: <a href="https://github.com/Classiq/classiq-library/blob/main/tutorials/phase_kickback/phase_kickback.ipynb" target="_blank">https://github.com/Classiq/classiq-library/blob/main/tutorials/phase_kickback/phase_kickback.ipynb</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
```
Comment (Nadav138) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>The link "here" seems to be broken</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (EdenSchirmanClassiq) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>Great openeing!</p><p> </p><p>As building an oracle is THE best example for the usefulness of high-level functional design, we want to emphasize it in the terminology. Please change the "construct an oracle in a quantum circuit" to "construct and design an oracle for a quantum algorithm?" or "construct and design an in practice"</p><p> </p><p>Then, I would explain a bit more the XOR operation and that the function is a binary function with a binary output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (EdenSchirmanClassiq) [inline tutorials/quantum_oracle/quantum_oracle_tutorial.ipynb:1]: <p>Again here: circuit-&gt; algorithm/model</p><p> </p><p>This pargraph should really emphasize the power of high-level functional design and that the circuit implementation of each oracle model can vary substantially according to the constraints and preferences. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Classiq/classiq-library/pull/186/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Comment by amir-naveh** [comment]
Hi @alexandre-ricardo , @EdenSchirmanClassiq - This has been inactive for a while, please re-open if relevant. Thank you

**Pull request closed without merging**